[stdlib] Hristo/foundations for os process module

izo0x90 opened the pull request:
Sets up the foundation for implementing the `os/process` module [PR with process module changes](https://github.com/modular/mojo/pull/3998)
  - Add `read_bytes` capability to `FileDecscriptor`
  - Add file descriptor controls function to Libc. bindings
  - Adds vfork, execvp, kill system call utils. to Mojos cLib binds

izo0x90 committed:
Adds `clib` functionality to spawn and manage processes from exec. file

  - Add `read_bytes` capability to `FileDecscriptor`
  - Add file descriptor controls function to Libc. bindings
  - Adds vfork, execvp, kill system call utils. to Mojos cLib binds

Signed-off-by: Hristo I. Gueorguiev <53634432+izo0x90@users.noreply.github.com>

izo0x90 committed:
Update file_descriptor `read_bytes` to return List of bytes

Update `read_bytes` from Span[Byte] to List[Byte] for better memory
safety and "user ergonomics"

Signed-off-by: Hristo I. Gueorguiev <53634432+izo0x90@users.noreply.github.com>

martinvuyk reviewed (commented):
Hi, just 2 comments :)

izo0x90 committed:
Update file_descriptor `read_bytes` internals

Update the list creation to conform to upcoming List API changes that
make the list size a `private attribute`

Signed-off-by: Hristo I. Gueorguiev <53634432+izo0x90@users.noreply.github.com>

ConnorGray reviewed (commented):
Hi @izo0x90, thanks for splitting this off, this is great! 🙂

I've left a few minor comments, mostly suggestions that we include links to the POSIX docs where appropriate. Once those are addressed, I'll proceed with merging this in.

In the meantime, I'll `!sync` this preemptively, just to ensure it passes internal CI.

Thanks again for contributing to Mojo 🔥

izo0x90 committed:
Update stdlib/src/sys/_libc.mojo

Co-authored-by: Connor Gray <accounts@connorgray.com>
Signed-off-by: Hristo (Izo) G. <53634432+izo0x90@users.noreply.github.com>

izo0x90 committed:
Update stdlib/src/sys/_libc.mojo

Co-authored-by: Connor Gray <accounts@connorgray.com>
Signed-off-by: Hristo (Izo) G. <53634432+izo0x90@users.noreply.github.com>

izo0x90 committed:
Update stdlib/src/sys/_libc.mojo

Co-authored-by: Connor Gray <accounts@connorgray.com>
Signed-off-by: Hristo (Izo) G. <53634432+izo0x90@users.noreply.github.com>

izo0x90 committed:
Update stdlib/src/sys/_libc.mojo

Co-authored-by: Connor Gray <accounts@connorgray.com>
Signed-off-by: Hristo (Izo) G. <53634432+izo0x90@users.noreply.github.com>

izo0x90 committed:
Update stdlib/src/sys/_libc.mojo

Co-authored-by: Connor Gray <accounts@connorgray.com>
Signed-off-by: Hristo (Izo) G. <53634432+izo0x90@users.noreply.github.com>

izo0x90 committed:
Update stdlib/src/sys/_libc.mojo

Co-authored-by: Connor Gray <accounts@connorgray.com>
Signed-off-by: Hristo (Izo) G. <53634432+izo0x90@users.noreply.github.com>

izo0x90 committed:
Update stdlib/src/sys/_libc.mojo

Co-authored-by: Connor Gray <accounts@connorgray.com>
Signed-off-by: Hristo (Izo) G. <53634432+izo0x90@users.noreply.github.com>

izo0x90 committed:
Update `file_descriptor.read_bytes` do not alloc.

- `read_bytes` takes in Span[Byte] buffer as opposed to allocating
  to avoid n+1 alloc. issues etc.

- Remove c_str_ptr type from ffi module since it is not a convention set
  by C lang., it will potentially get added back in at a later date
  after discussions and under its own PR

- Doc string updates to libC funcs.

Co-authored-by: Connor Gray <accounts@connorgray.com>
Signed-off-by: Hristo (Izo) G. <53634432+izo0x90@users.noreply.github.com>

martinvuyk reviewed (commented):
LGTM

izo0x90 committed:
Update stdlib/src/builtin/file_descriptor.mojo

Co-authored-by: martinvuyk <110240700+martinvuyk@users.noreply.github.com>
Signed-off-by: Hristo (Izo) G. <53634432+izo0x90@users.noreply.github.com>

izo0x90 committed:
Update stdlib/src/builtin/file_descriptor.mojo

Co-authored-by: martinvuyk <110240700+martinvuyk@users.noreply.github.com>
Signed-off-by: Hristo (Izo) G. <53634432+izo0x90@users.noreply.github.com>

izo0x90 committed:
Apply suggestions from code review

Co-authored-by: martinvuyk <110240700+martinvuyk@users.noreply.github.com>
Signed-off-by: Hristo (Izo) G. <53634432+izo0x90@users.noreply.github.com>
